pkg/cache/sql/informer/listoption_indexer.go

```diff
@@ -405,37 +467,69 @@ func (l *ListOptionIndexer) validateColumn(column string) error {
 
 // buildORClause creates an SQLite compatible query that ORs conditions built from passed filters
 func (l *ListOptionIndexer) buildORClauseFromFilters(orFilters OrFilter) (string, []any, error) {
-	var orWhereClause string
 	var params []any
+	clauses := make([]string, 0, len(orFilters.Filters))
+	var newParams []any
+	var newClause string
+	var err error
 
-	for index, filter := range orFilters.Filters {
-		opString := "LIKE"
-		if filter.Op == NotEq {
-			opString = "NOT LIKE"
+	for _, filter := range orFilters.Filters {
+		if isLabelFilter(&filter) {
+			newClause, newParams, err = l.getLabelFilter(filter)
+		} else {
+			newClause, newParams, err = l.getFieldFilter(filter)
 		}
-		columnName := toColumnName(filter.Field)
-		if err := l.validateColumn(columnName); err != nil {
+		if err != nil {
 			return "", nil, err
 		}
+		clauses = append(clauses, newClause)
+		params = append(params, newParams...)
+	}
+	return strings.Join(clauses, " OR "), params, nil
+}
 
-		orWhereClause += fmt.Sprintf(`f."%s" %s ? ESCAPE '\'`, columnName, opString)
-		format := strictMatchFmt
-		if filter.Partial {
-			format = matchFmt
-		}
-		match := filter.Match
-		// To allow matches on the backslash itself, the character needs to be replaced first.
-		// Otherwise, it will undo the following replacements.
-		match = strings.ReplaceAll(match, `\`, `\\`)
-		match = strings.ReplaceAll(match, `_`, `\_`)
-		match = strings.ReplaceAll(match, `%`, `\%`)
-		params = append(params, fmt.Sprintf(format, match))
-		if index == len(orFilters.Filters)-1 {
-			continue
-		}
-		orWhereClause += " OR "
+func (l *ListOptionIndexer) getFieldFilter(filter Filter) (string, []any, error) {
+	opString := "LIKE"
+	if filter.Op == NotEq {
+		opString = "NOT LIKE"
+	}
+	columnName := toColumnName(filter.Field)
+	if err := l.validateColumn(columnName); err != nil {
+		return "", nil, err
+	}
+
+	clause := fmt.Sprintf(`f."%s" %s ? ESCAPE '\'`, columnName, opString)
+	return clause, []any{formatMatchTarget(filter)}, nil
+}
+
+func (l *ListOptionIndexer) getLabelFilter(filter Filter) (string, []any, error) {
+	opString := "LIKE"
+	if filter.Op == NotEq {
+		opString = "NOT LIKE"
+	}
```

Comment:
I fear this will not suffice.

Per the "Proposed solution" paragraph in [the issue](https://github.com/rancher/rancher/issues/46333#issue-2429999150) we'd really need the full label selector syntax and power, per https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors - with some kind of parser on the Steve side.

Reply:
Thanks ... I missed that note. Yes, I'll add a proper op parser.

Reply:
Do we really need to implement the entire Label Selector command set from kubectl? Because here isn't the only client the UI, and which operators does it need?

^ @richard-cox

Reply:
It's easy enough to call `(k8s.io/apimachinery/pkg/labels/selector.go) labels.Parse()` or `labels.ParseToRequirements()` to pass in a standard k8s label selector expression and get back a list of requirements, but I'm not convinced it's needed.

Reply:
Unfortunately full label selector support is needed (second point in https://github.com/rancher/rancher/issues/46333#issuecomment-2250742787). The UI has a really useful feature used in a number of places to show the result of label selector config. At the moment it is done locally on all resources, which is a scaling nightmare. It needs to change to on-demand via the new API.

Reply:
@ericpromislow good that we can have the parsing done easily

Reply:
I need to see if there are any conflicts between the current filter language and the kubectl label selection language.

Reply:
@moio @richard-cox 

So there are several problems with using the kubernetes label parser.

First, it requires values to match (the right hand part of '=' or another match operator) the
k8s label syntax as described at https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set 

Second, quotations aren't allowed in the target values, so we can't rely on the label parser to
determine whether to do an exact match or not.

It might make sense to do label parsing only for these sub-queries:

`metadata.labels.NAME <REST-OF-SUBQUERY>`
and
`metadata.labels.[IDENTIFIER] <REST-OF-SUBQUERY>`

and then we have to preprocess the expression a bit into the format
`<NAME | IDENTIFIER> <REST-OF-SUBQUERY>`

Now if you want to support all the k8s label-parser operators:

```
=
!=
EXISTS
! EXISTS
IN (set)
NOTIN (set)
```

we need to distinguish commas inside sets from commas that act like an or-operator.
I've pushed a failing test to https://github.com/ericpromislow/steve/blob/50dbbdd4000e31051123856729aee7075d19d480/pkg/stores/sqlpartition/listprocessor/processor_test.go#L269-L295

We also need to add quotation marks as delimiters to indicate exact match.

The ad-hoc parser doesn't handle the existence and set-membership tests, so I'm going to copy the
k8s parser (Apache-licensed) into our code and modify it. I'm not worried about forking it -- the label
selection syntax is unlikely to change in the near term, and if it does it shouldn't be difficult to incorporate the changes.

Reply:
Would you be able to provide some sample url params demonstrating how the label selector would work?

UI side i don't think we're tied to a specific format, we would need to convert the json representation of the label selector to params strings in all worlds. Does that help at all?

Reply:
@ericpromislow how about making `labelSelector` a standalone filter query parameter, on par with `filter`, `sort`, `limit`, etc?

If we do so, commas retain their original Steve meaning in the scope of `filter` (`OR` between filters) and the Kubernetes standard meaning in the scope of `labelSelector` (`AND` between selectors or set-element separator - it'll be up to the labelSelector parsing code to figure that out). It will be implied that any `labelSelector` acts on `metadata.labels` - it's implicit, like everywhere else in Kubernetes.

Rewording my question: could we add a top-level `labelSelector` parameter and then treat its value as a black box passing it straight to Kubernetes parsing code and not worry about anything else?

Reply:
References:
 - https://github.com/rancher/steve?tab=readme-ov-file#list-specific-query-parameters
 - https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors

Reply:
FWIW the UI will be taking selectors config from https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.28/#labelselector-v1-meta and making a request to get the result

labelSelectors contain two parts
- array of matchExpression. basically `key` - `operator` (in, not in, less than, etc), - `values`
- matchLabel is shorthand for `key` - `IN` - `values`

Reply:
There's only so much of a reason to use the k8s label selectors as is, and that's because we aren't using the k8s matching mechanism (also in apimachinery), but converting the parsed queries into SQL queries. I'm doing that now with the forked parser.

The only mod I needed to make to the label query parser was support for single-quoted and double-quoted strings. And we get errors on unterminated strings basically for free (including `...\\` at the end of input).

I believe I've correctly implemented everything for https://github.com/rancher/steve?tab=readme-ov-file#filter

I haven't implemented support for less-than and greater-than yet (the k8s language predictably uses "<" and ">"). Right now they're parsed and the code that converts a k8s `Requirement` into an `informer.Filter` throws an error when it sees one of those. Let me go build a downstream cluster so I can see what happens with selecting `provisioning.cattle.io/v1/Cluster`s

Reply:
> Note that for non-identifier target values, they need to be quoted, so you can't do a LIKE match on a target value like we contain spaces. You can do a LIKE match on one of its words, or a subword, but not anything more complex than a single identifier-type token.

@richard-cox is this acceptable from your perspective?

In particular, where does the UI get those label selectors from? Are they from user input, are they constructed by UI code from user input, or do they come from other Kubernetes objects that have labelSelectors in them, such as these https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements?

IOW how close of a syntax match do you need. Does it really have to match 1:1 100% with the Kubernetes spec?

Reply:
> In particular, where does the UI get those label selectors from? Are they from user input, are they constructed by UI code from user input, or do they come from other Kubernetes objects that have labelSelectors in them

All of the above. Use cases are listed in https://github.com/rancher/dashboard/issues/10417#issue-2123233258 but very generally

1. User is creating a new labelSelector object via input fields and the UI will show the result of those proposed selectors (it will then persist the changes as a labelSelector object within the resource)
2. Applying the labelSelector from within a resource to find other resources

So the UI isn't using label selector like functionality, it's literally label selector stuff. Being able to match the exact functionality is important. It would be a collection of [these](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.30/#labelselectorrequirement-v1-meta) AND'd together

key: string
operator: In, NotIn, Exists and DoesNotExist
values: string array

> Note there is also [Node labelSelector](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.30/#labelselectorrequirement-v1-meta) requirements that contain Gt (greater than) and Lt (less than)

Examples
```
foo IN [bar]
foo IN [bar, baz2]
aaa NotIn [bar, baz2]
bbb Exists
ccc DoesNotExist
ddd Gt 1
eee Lt 2
```

which i think equates to 

```
  labelSelector
    matchLabels:
     foo: bar
    matchExpressions:
     - key: foo
       op: In
       values: [bar, baz]
     - key: aaa
       op: NotIn
       values: [bar, baz]
     - key: bbb
       op: Exists
     - key: ccc
       op: DoesNotExist
     - key: ddd
       op: Gt
       values: 1
     - key: eee
       op: Lt
       values: 2
```

@ericpromislow thanks for providing an example below (`filter=metadata.labels%5bauthz.management.cattle.io/default-project%5d="true"'`) can you provide filter param equivalents for the examples above?

> can't do a LIKE match on a target value like we contain spaces. You can do a LIKE match on one of its words, or a subword, but not anything more complex than a single identifier-type token.

I couldn't find LIKE functionality in labelSelector land, so as long as it doesn't break existing functionality i think that's fine

Reply:
@richard-cox  Here how you would encode those above examples:
```
foo IN [bar] => ?filter=foo+IN+(bar)
foo IN [bar, baz2] => ?filter=foo+IN+(bar,baz2) 
aaa NotIn [bar, baz2]=> ?filter=foo+NOTIN+(bar,baz2) 
bbb Exists=> ?filter=bbb
ccc DoesNotExist ?filter=!bbb. # or %21bbb
```

General comments:
1. The case of the keywords is always ignore, so something like `eXiSt` is ok, but I use ALLCAPS for clarity.
2. White space is ignored outside strings.  The second request could equivalently be
`?filter=foo+IN+++(+bar+,++++++baz2++++)`

3. The `EXISTS` and `NOT EXISTS` implementations don't make a lot of sense in terms of what gets returned. Right now `EXISTS` returns the row if it's matched, and in fact I can't get `NOT EXISTS` to work.

But I think it would make more sense if the two existence tests just return something simpler than a row-set ... I'll need to work on that.

The next two haven't been implemented yet, but should be:
```
ddd Gt 1 => ?filter=ddd+<+1
eee Lt 2=> ?filter=eee+>+2
```

Reply:
As per slack message, this approach i think should work but would like to prod it with some real life queries.. so awaiting a dev image